mock_drivers/rosbag_mock_drivers/include/rosbag_mock_drivers/MockDriverNode.h

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2019-2020 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+#pragma once
+
+#include <ros/ros.h>
+#include <carma_utils/CARMAUtils.h>
+#include "comm_types.h"
+#include "ROSComms.h"
+#include <vector>
+
+namespace mock_drivers
+{
+/// \brief Wrapper for publishers which allow their base topic name to be extracted
+struct PublisherWapper
+{
+  ros::Publisher pub;
+  std::string base_topic_name;
+};
+
+/// \brief Node class which connects to the ROS network
+class MockDriverNode
+{
+private:
+  boost::shared_ptr<ros::CARMANodeHandle> cnh_;
+  std::vector<PublisherWapper> publishers_;
+  std::vector<ros::Subscriber> subscribers_;
+  std::vector<ros::ServiceServer> services_;
+
+  // used for testing
+  bool dummy_ = false;  // if the mock driver node is a dummy then it wont initialize a CARMANodeHandle and its functions wont
+                // do anything
+  std::vector<std::string> topics_;
+  std::vector<ros::Time> time_stamps_;
+
+public:
+
+  /**
+   * \brief Returns the fully qualified name of this node in the ROS network graph
+   * 
+   * \return The fully specified name
+   */ 
+  std::string getGraphName() const;
+
+  /*! \brief Function to add a publisher from a ROSComms object */
+  template <typename T>
+  void addPub(T comm)
+  {
+    if (!dummy_)
+    {
+      PublisherWapper pub;
+      pub.pub = cnh_->advertise<decltype(comm->getTemplateType())>(comm->getTopic(), comm->getQueueSize());
+      pub.base_topic_name = comm->getTopic();
+      publishers_.push_back(pub);
+    }
+  }
+
+  /*! \brief Function to add a subscriber from a ROSComms object */
+  template <typename T>
+  void addSub(T comm)
+  {
+    if (!dummy_)
+    {
+      subscribers_.push_back(cnh_->subscribe<decltype(comm->getTemplateType())>(
+          comm->getTopic(), comm->getQueueSize(), &ROSComms<decltype(comm->getTemplateType())>::callback, comm));
+    }
+  }
+
+  /*! \brief Function to add a service from a ROSComms object */
+  template <typename T>
+  void addSrv(T comm)
+  {
+    if (!dummy_)
+    {
+      services_.push_back(cnh_->advertiseService(
+          comm->getTopic(), &ROSComms<decltype(comm->getReqType()), decltype(comm->getResType())>::callback, comm));
+    }
+  }
+
+  // TODO (needs at least C++ 17): use a constexpr if statement to an addComms function so you don't need different add
```

Comment:
Do we need to keep this for future reference?

Reply:
No I suppose not. We can always look for it in the git history if needed. I'll delete it.

Reply:
Done